lib/lua_lpm.c

```diff
@@ -0,0 +1,409 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <arpa/inet.h>
+#include <lauxlib.h>
+#include <sys/time.h>
+
+#include <rte_lcore.h>
+
+#include "luajit-ffi-cdata.h"
+#include "gatekeeper_net.h"
+#include "gatekeeper_lpm.h"
+#include "gatekeeper_lua_lpm.h"
+
+#define CTYPE_STRUCT_LPM_PTR "struct rte_lpm *"
+
+int
+l_init_lpm(lua_State *l)
+{
+	struct rte_lpm_config lpm_conf;
+	struct timeval now;
+	uint32_t correct_ctypeid_lpm = luaL_get_ctypeid(l,
+		CTYPE_STRUCT_LPM_PTR);
+	void *lpm_cdata;
+
+	/* First argument must be a Lua number. */
+	luaL_checktype(l, 1, LUA_TNUMBER);
+
+	/* Second argument must be a Lua number. */
+	luaL_checktype(l, 2, LUA_TNUMBER);
+
+	if (lua_gettop(l) != 2)
+		luaL_error(l, "Expected two arguments, however it got %d arguments",
+			lua_gettop(l));
+
+	lpm_conf.max_rules = lua_tointeger(l, 1);
+	lpm_conf.number_tbl8s = lua_tointeger(l, 2);
+
+	gettimeofday(&now, NULL);
+
+	lpm_cdata = luaL_pushcdata(l, correct_ctypeid_lpm,
+		sizeof(struct rte_lpm *));
+	*(struct rte_lpm **)lpm_cdata = init_ipv4_lpm("gt_policy", &lpm_conf,
+		rte_socket_id(), rte_lcore_id(),
+		now.tv_sec * 1000000 + now.tv_usec);
+
+	lua_remove(l, 1);
+	lua_remove(l, 1);
+
+	return 1;
+}
+
+int
+l_lpm_add(lua_State *l)
+{
+	int ret;
+	int ip_type;
+	size_t len;
+	const char *ip_addr;
+	uint8_t depth;
+	uint32_t policy_id;
+	struct in_addr ipv4_addr;
+	struct rte_lpm *lpm;
+	uint32_t ctypeid;
+	uint32_t correct_ctypeid_lpm = luaL_get_ctypeid(l,
+		CTYPE_STRUCT_LPM_PTR);
+
+	/* First argument must be of type CTYPE_STRUCT_LPM_PTR. */
+	void *cdata = luaL_checkcdata(l, 1, &ctypeid, CTYPE_STRUCT_LPM_PTR);
+	if (ctypeid != correct_ctypeid_lpm)
+		luaL_error(l, "Expected `%s' as first argument",
+			CTYPE_STRUCT_LPM_PTR);
+
+	/* Second argument must be a Lua string. */
+	luaL_checktype(l, 2, LUA_TSTRING);
+
+	/* Third argument must be a Lua number. */
+	luaL_checktype(l, 3, LUA_TNUMBER);
+
+	/* Fourth argument must be a Lua number. */
+	luaL_checktype(l, 4, LUA_TNUMBER);
+
+	if (lua_gettop(l) != 4)
+		luaL_error(l, "Expected four arguments, however it got %d arguments",
+			lua_gettop(l));
+
+	lpm = *(struct rte_lpm **)cdata;
+	ip_addr = lua_tolstring(l, 2, &len);
```

Comment:
Require the IPv4 address as a number as `lpm_del()` does to keep the interface symmetric.

This comment applies to the IPv6 case as well.